### ISSUE TITLE
Rsync via tar strategy: Cannot open: Not a directory ERROR

### ISSUE THREAD

**Issue by bdshadow** [openshift/origin#14075]
`oc rsync` via tar strategy is buggy

##### Version
oc v1.4.1+3f9807a
kubernetes v1.4.0+776c994

##### Environment
Windows (must not be important)
`rsync` is deleted, only `tar` is available

##### Steps To Reproduce
1. create eap application in openshift, e.g. jboss-eap70-openshift
2.  wait till pod is built and run
3a. run rsync from pod to local
`oc rsync --token=ToKeN--server=https://10.1.2.2:8443 --insecure-skip-tls-verify=true --exclude='.git' jboss-eap70-openshift-1-zn39w:/opt/eap/standalone/deployments/ -n myproject C:\Users\dbocharo\dev\runtime-New_configuration\.metadata\.plugins\org.jboss.ide.eclipse.as.core\myproject@jboss-eap70-openshift\deploy\`
3b. run rsync from local to pod
`oc rsync --token=ToKen --server=https://10.1.2.2:8443 --insecure-skip-tls-verify=true --exclude='.git' C:\Users\dbocharo\dev\runtime-New_configuration\.metadata\.plugins\org.jboss.ide.eclipse.as.core\myproject@jboss-eap70-openshift\deploy\ jboss-eap70-openshift-1-zn39w:/opt/eap/standalone/deployments/ -n myproject`

##### Current Result
3a. 
```bash
WARNING: rsync command not found in path. Download cwRsync for Windows and add it to your PATH.
./ROOT.war
ERROR: Error extracting file "./ROOT.war": open C:\Users\dbocharo\dev\runtime-New_configuration\.metadata\.plugins\org.jboss.ide.eclipse.as.core\myproject@jboss-eap70-openshift\deploy\ROOT.war: is a directory
ERROR: Error extracting tar stream
Ignoring the following flags because they only apply to rsync: --exclude
error: error extracting tar at destination directory: open C:\Users\dbocharo\dev\runtime-New_configuration\.metadata\.plugins\org.jboss.ide.eclipse.as.core\myproject@jboss-eap70-openshift\deploy\ROOT.war: is a directory
```
3b.
```bash
WARNING: rsync command not found in path. Download cwRsync for Windows and add it to your PATH.
README.txt
ROOT.war/META-INF/MANIFEST.MF
ROOT.war/META-INF/maven/org.jboss.quickstarts.eap/jboss-kitchensink/pom.properties
ROOT.war/META-INF/maven/org.jboss.quickstarts.eap/jboss-kitchensink/pom.xml
ROOT.war/WEB-INF/beans.xml
ROOT.war/WEB-INF/classes/META-INF/persistence.xml
ROOT.war/WEB-INF/classes/import.sql
ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/controller/MemberController.class
ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/data/MemberListProducer.class
ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/data/MemberRepository.class
ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/model/Member.class
ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/rest/JaxRsActivator.class
ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/rest/MemberResourceRESTService.class
ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/service/MemberRegistration.class
ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/util/Resources.class
ROOT.war/WEB-INF/faces-config.xml
ROOT.war/WEB-INF/kitchensink-quickstart-ds.xml
ROOT.war/WEB-INF/templates/default.xhtml
ROOT.war/index.html
ROOT.war/index.xhtml
ROOT.war/resources/css/screen.css
ROOT.war/resources/gfx/asidebkg.png
ROOT.war/resources/gfx/banner.png
ROOT.war/resources/gfx/bkg-blkheader.png
ROOT.war/resources/gfx/headerbkg.png
ROOT.war/resources/gfx/rhjb_eap_logo.png
ROOT.war.dodeploy
ROOT.war.isdeploying
activemq-rar.rar
activemq-rar.rar.isdeploying
Ignoring the following flags because they only apply to rsync: --exclude
tar: ROOT.war/META-INF/MANIFEST.MF: Cannot open: Not a directory
tar: ROOT.war/META-INF/maven/org.jboss.quickstarts.eap/jboss-kitchensink/pom.properties: Cannot open: Not a directory
tar: ROOT.war/META-INF/maven/org.jboss.quickstarts.eap/jboss-kitchensink/pom.xml: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/beans.xml: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/classes/META-INF/persistence.xml: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/classes/import.sql: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/controller/MemberController.class: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/data/MemberListProducer.class: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/data/MemberRepository.class: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/model/Member.class: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/rest/JaxRsActivator.class: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/rest/MemberResourceRESTService.class: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/service/MemberRegistration.class: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/classes/org/jboss/as/quickstarts/kitchensink/util/Resources.class: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/faces-config.xml: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/kitchensink-quickstart-ds.xml: Cannot open: Not a directory
tar: ROOT.war/WEB-INF/templates/default.xhtml: Cannot open: Not a directory
tar: ROOT.war/index.html: Cannot open: Not a directory
tar: ROOT.war/index.xhtml: Cannot open: Not a directory
tar: ROOT.war/resources/css/screen.css: Cannot open: Not a directory
tar: ROOT.war/resources/gfx/asidebkg.png: Cannot open: Not a directory
tar: ROOT.war/resources/gfx/banner.png: Cannot open: Not a directory
tar: ROOT.war/resources/gfx/bkg-blkheader.png: Cannot open: Not a directory
tar: ROOT.war/resources/gfx/headerbkg.png: Cannot open: Not a directory
tar: ROOT.war/resources/gfx/rhjb_eap_logo.png: Cannot open: Not a directory
tar: ROOT.war.dodeploy: time stamp 2017-05-05 14:09:19 is 77342.946681899 s in the future
tar: Exiting with failure status due to previous errors
error: error extracting tar at destination directory: command terminated with exit code 2
```
##### Expected Result
rsync is successful

**Comment by apupier** [comment]
do we have some news on this item? it was reported in May and there was no updates. Is it because it requires a lot of work? Or just a low priority?

the downside of this issue is that rsync is useful during development process. With tar mode not working properly, as far as I understand, it is forcing to modify the images deployed during development to include rsync which seems a bit hard to manage a different image for development and runtime.

**Comment by juanvallejo** [comment]
@bdshadow any chance you could provide the app template you are using? Will try to replicate bug. Are you seeing this only on Windows?

**Comment by juanvallejo** [comment]
@fabianofranz 
Looks like for the output in `3a`, the error is happening in [this method in source-to-image](https://github.com/openshift/source-to-image/blob/master/pkg/tar/tar.go#L330). It is receiving the file header for `ROOT.war` from the tar stream and, I assume, it is unable to figure out whether it's a file or directory (maybe file mode info missing?), and defaulting to treating it like a file. Not sure what a good fix for this might be, maybe a better error message? We could do:

```go
switch header.TypeFlag {
case tar.TypeDir:
    // current logic for handling dir
case tar.TypeReg:
    // current logic for handling files
default:
  // error stating that file type could not be determined for file header.Name
}
```

in the source-to-image `ExtractTarStreamFromTarReader` method.

For the output in 3b, it just looks like it fails for the opposite reason: every file inside of `ROOT.war` seems to have a directory attribute.

I think the problem here might just lie with the file attributes for `ROOT.war`.
@apupier @bdshadow have you seen this issue occur with any other files / directories?

**Comment by fabianofranz** [comment]
I was unable to reproduce this on 3.6.x using a similar environment: on Windows, without rsync installed, with a ROOT.war directory as part of the synced structured, eap pod, both to and from pod.

Since it's unlikely the fix will be ported to 1.4 I'm closing this, but feel free to reopen with more information about how to reproduce it if it still happens on a recent (3.5+) version.
